### PR TITLE
Feat/scheduled events

### LINKED ISSUE #258: Add guild scheduled events endpoints, gateway events, audit log, invite, and permissions changes.
### Summary

Add something that has been documented in ddocs

### What is the feature request for?

The core library

### The Problem

https://github.com/discord/discord-api-docs/commit/1298152172d3359abf069d1b24d4bbc85532b735

### The Ideal Solution

do it

### The Current Solution

not doing it

### Additional Context

https://canary.discord.com/channels/881118111967883295/881124672056467456/909890698873032745

### PR DESCRIPTION
**Pull request by ooliver1**
## Summary

closes #258 

- guild attributes and methods
- intent flag
- dispatch events
- relevant enums and classes
- docs

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

I have tested to some extent but maybe others to test it for different setups would be cool

### CODE REVIEW

**Comment by veni-vidi-code** [comment]
I approve Codeowners change, had no time yet to read rest

File: docs/api.rst [hunk 1]
```diff
@@ -2608,6 +2608,50 @@ of :class:`enum.Enum`.
 
         The guild may contain NSFW content.
 
+.. class:: EntityType
```
Comment (TAG-Epic) [inline docs/api.rst:2611]: This should probably start with ScheduledEvent or does it not do that in the docs?

File: nextcord/__init__.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 A basic wrapper for the Discord API.
 
-:copyright: (c) 2015-present Rapptz
+:copyright: (c) 2015-present Rapptz and 2021-present tag-epic
```
Comment (TAG-Epic) [inline nextcord/__init__.py:7]: We prefer using a second line. See enums file

File: nextcord/enums.py [hunk 3]
```diff
@@ -589,6 +601,23 @@ class NSFWLevel(Enum, comparable=True):
     age_restricted = 3
 
 
+class EntityType(Enum):
```
Comment (TAG-Epic) [inline nextcord/enums.py:604]: Same here?

File: nextcord/ext/commands/__init__.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 An extension module to facilitate creation of bot commands.
 
-:copyright: (c) 2015-present Rapptz
+:copyright: (c) 2015-present Rapptz and 2021-present tag-epic
```
Comment (TAG-Epic) [inline nextcord/ext/commands/__init__.py:7]: Same here

**Review by TAG-Epic — CHANGES_REQUESTED** [review]

**Comment by ooliver1** [comment]
1 sec forgot to commit smth

File: nextcord/guild.py [hunk 5]
```diff
@@ -2966,3 +2996,161 @@ async def change_voice_state(
...
+                print(event.name)
+
+        Flattening into a list ::
+
+            events = await guild.fetch_scheduled_events().flatten()
+            # events is now a list of ScheduledEvent...
+        """
+        return ScheduledEventIterator(self, with_users=with_users)
+
+    def get_scheduled_event(self, event_id: int) -> Optional[ScheduledEvent]:
```
Comment (Skelmis) [inline nextcord/guild.py:3045]: Are these ``Snowflake``? I'm not sure but want to ask

File: nextcord/guild.py [hunk 6]
```diff
@@ -2966,3 +2996,161 @@ async def change_voice_state(
...
+        Returns
+        -------
+        Optional[ScheduledEvent]
+            The event object, if found.
+        """
+        return self._scheduled_events.get(event_id)
+
+    async def fetch_scheduled_event(
+        self,
+        event_id,
```
Comment (Skelmis) [inline nextcord/guild.py:3067]: Type `event_id`

File: nextcord/scheduled_events.py [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+                try:
+                    self.member: Optional[Member] = self.guild.get_member(
+                        member['id']
+                    )
+                except KeyError:
+                    m = Member(data=member, guild=self.guild, state=self._state)
+                    self.member: Optional[Member] = m
+            else:
+                m = Member(data=member, guild=self.guild, state=self._state)
+                self.member: Optional[Member] = m
```
Comment (Skelmis) [inline nextcord/scheduled_events.py:116]: Is there a need for making this 2 lines rather then 1?

File: nextcord/state.py [hunk 8]
```diff
@@ -1405,6 +1408,44 @@ def create_message(
     ) -> Message:
         return Message(state=self, channel=channel, data=data)
 
+    def create_scheduled_event(
```
Comment (Skelmis) [inline nextcord/state.py:1411]: There's a discrepancy between files. Some refer to them as `scheduled_event`, then others are just `guild_event` ..?

**Review by Skelmis — CHANGES_REQUESTED** [review]

File: nextcord/scheduled_events.py (hunk 7 above)
Comment (ooliver1) [reply nextcord/scheduled_events.py:116]: i think it takes up too much line length

File: nextcord/state.py (hunk 8 above)
Comment (ooliver1) [reply nextcord/state.py:1411]: could you show me where its called guild event?

File: nextcord/guild.py (hunk 5 above)
Comment (ooliver1) [reply nextcord/guild.py:3045]: get functions use ints as they search in the dictionary ``Dict[int, obj]``, for consistency with other get_x methods

File: nextcord/state.py (hunk 8 above)
Comment (ooliver1) [reply nextcord/state.py:1411]: edited guild methods

**Review by RPSMain — APPROVED** [review]

**Comment by ooliver1** [comment]
waiting for reviews on this

**Comment by ooliver1** [comment]
user add and remove events added

File: nextcord/client.py [hunk 9]
```diff
@@ -862,6 +863,23 @@ def get_sticker(self, id: int, /) -> Optional[GuildSticker]:
...
 
+    def get_scheduled_event(self, id: int, /) -> Optional[ScheduledEvent]:
+        """Returns a scheduled event with the given ID.
+
+        .. versionadded:: 2.0
+
+        Parameters
+        -----------
+        id: :class:`int`
+            The ID to search for.
```
Comment (Skelmis) [inline nextcord/client.py:874]: "The scheduled event's ID to search for." or somethin

File: nextcord/client.py [hunk 10]
```diff
@@ -1643,3 +1661,14 @@ def persistent_views(self) -> Sequence[View]:
         .. versionadded:: 2.0
         """
         return self._connection.persistent_views
+
+    @property
+    def scheduled_events(self) -> List[ScheduledEvent]:
+        """List[ScheduledEvent]: A list of scheduled events
```
Comment (Skelmis) [inline nextcord/client.py:1667]: Why isn't this denoted in the docstring guide as returns

```
Returns
--------
List[ScheduledEvent]
 A list of scheduled events
```

File: nextcord/ext/commands/errors.py [hunk 11]
```diff
@@ -464,6 +465,24 @@ def __init__(self, argument: str) -> None:
         self.argument: str = argument
         super().__init__(f'Sticker "{argument}" not found.')
 
+
+class ScheduledEventNotFound(BadArgument):
```
Comment (Skelmis) [inline nextcord/ext/commands/errors.py:469]: Is this in the docs?

File: nextcord/guild.py [hunk 12]
```diff
@@ -572,6 +594,14 @@ def categories(self) -> List[CategoryChannel]:
         r.sort(key=lambda c: (c.position, c.id))
         return r
 
+    @property
+    def scheduled_events(self) -> List[ScheduledEvent]:
```
Comment (Skelmis) [inline nextcord/guild.py:598]:
```Returns
--------
List[ScheduledEvent]
 A list of scheduled events
```

File: nextcord/guild.py [hunk 13]
```diff
@@ -2966,3 +2996,161 @@ async def change_voice_state(
...
+
+            This method is an API call. For general usage, consider 
+            :attr:`scheduled_events` instead.
+
+        .. versionadded:: 2.0
+
+        Parameters
+        ----------
+        with_users: Optional[:class:`bool`]
+            If the event should be received with :attr:`ScheduledEvent.users`
```
Comment (Skelmis) [inline nextcord/guild.py:3018]: Discuss the default and what it implies

File: nextcord/guild.py [hunk 14]
```diff
@@ -2966,3 +2996,161 @@ async def change_voice_state(
...
+        if start_time is not MISSING:
+            payload['scheduled_start_time'] = start_time.isoformat()
+        if end_time is not MISSING:
+            payload['scheduled_end_time'] = end_time.isoformat()
+        if description is not MISSING:
+            payload['description'] = description
+        if entity_type is not MISSING:
+            payload['entity_type'] = entity_type.value
+        data = await self._state.http.create_event(self.id, **payload)
+        return self._store_scheduled_event(data)
```
Comment (Skelmis) [inline nextcord/guild.py:3156]: The code denotes that most of these as optional, why not document them as such?

File: nextcord/state.py [hunk 15]
```diff
@@ -1400,11 +1403,99 @@ def get_channel(self, id: Optional[int]) -> Optional[Union[Channel, Thread]]:
             if channel is not None:
                 return channel
 
+    def get_scheduled_event(self, id: Optional[int]) -> Optional[ScheduledEvent]:
+        if id is None:
```
Comment (Skelmis) [inline nextcord/state.py:1407]: It's only typed as none, not defaulted to None so this wouldn't occur?

**Review by Skelmis — CHANGES_REQUESTED** [review]

File: nextcord/client.py (hunk 9 above)
Comment (ooliver1) [reply nextcord/client.py:874]: all the other get_x do this, but ill follow this

File: nextcord/client.py (hunk 10 above)
Comment (ooliver1) [reply nextcord/client.py:1667]: properties can be done like this, see persistent_views for example

File: nextcord/ext/commands/errors.py (hunk 11 above)
Comment (ooliver1) [reply nextcord/ext/commands/errors.py:469]: thats a good point

File: nextcord/guild.py (hunk 12 above)
Comment (ooliver1) [reply nextcord/guild.py:598]: also properties can work like this

File: nextcord/guild.py (hunk 14 above)
Comment (ooliver1) [reply nextcord/guild.py:3156]: forgot to update that

File: nextcord/ext/commands/converter.py [hunk 16]
```diff
@@ -863,6 +864,74 @@ async def convert(self, ctx: Context, argument: str) -> nextcord.GuildSticker:
...
+        result = None
+        bot = ctx.bot
+        guild = ctx.guild
+
+        if match is None:
+            # Try to get the scheduled event by name. Try local guild first.
+            if guild:
+                result = nextcord.utils.get(guild.scheduled_events, name=argument)
+
+            if result is None:
```
Comment (TAG-Epic) [inline nextcord/ext/commands/converter.py:906]: This should be removed as event names are not unique.

**Review by TAG-Epic — CHANGES_REQUESTED** [review]

**Review by Lulalaby — APPROVED** [review]
LGTM Discord API Side.

File: nextcord/ext/commands/converter.py (hunk 16 above)
Comment (ooliver1) [reply nextcord/ext/commands/converter.py:908]: eh other converters like guild use names

Comment (TAG-Epic) [reply nextcord/ext/commands/converter.py:908]: True

**Comment by TAG-Epic** [comment]
Merge conflicts

**Review by vcokltfre — APPROVED** [review]
All seems to be in order, ready to merge!

**Pull request merged**